File: euler16.py
# Какова сумма цифр числа 2 в степени 1000?

from datetime import datetime


def euler16() -> None:
    """
    Решение задачи Эйлера №16.

    Какова сумма цифр числа 2 в степени 1000?
    """
    n = 1_000

    # 1 вариант
    start_time = datetime.now()
    print(sum(int(x) for x in str(2 ** n)))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    print(sum(map(int, str(2 ** n))))
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler16()

File: go/euler16.go
// euler16 находит сумму цифр числа 2 в степени 1000.
//
// Решение задачи Эйлера №16.
package main

import (
	"fmt"
	"math/big"
	"time"
)

func euler16() {
	n := 1000

	startTime := time.Now()
	result := big.NewInt(1)
	exp := big.NewInt(int64(n))
	base := big.NewInt(2)
	result.Exp(base, exp, nil)
	sum := 0
	for _, c := range result.String() {
		sum += int(c - '0')
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	result = big.NewInt(1)
	result.Exp(base, exp, nil)
	sum = 0
	for _, c := range result.String() {
		sum += int(c - '0')
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler16()
}
